### PR TITLE
feat(mark played): add ability to mark items as played or unplayed from card

### PR DESCRIPTION
**Pull request by Camc314**

### CODE REVIEW

File: components/Buttons/MarkWatchedButton.vue [hunk 1]
```diff
@@ -0,0 +1,72 @@
+<template>
+  <v-btn v-if="canMarkWatched" icon dark @click.stop.prevent="togglePlayed">
```
Comment (ThibaultNocchi) [inline components/Buttons/MarkWatchedButton.vue:2]: Why the `.stop.prevent`?

File: components/Buttons/MarkWatchedButton.vue [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  methods: {
+    ...mapActions('snackbar', ['pushSnackbarMessage']),
+    async togglePlayed(): Promise<void> {
+      try {
+        if (this.isPlayed) {
+          await this.$api.playState.markUnplayedItem({
+            userId: this.$auth.user.Id,
+            itemId: this.item.Id || ''
+          });
+          this.isPlayed = false;
```
Comment (ThibaultNocchi) [inline components/Buttons/MarkWatchedButton.vue:55]: I'd set `this.isPlayed` before the await call, and revert it in the catch. It'll be smoother for the user

**Comment by neopc10** [comment]
Can this be extended to music albums? Being able to filter out several thousand albums that I have heard or haven't heard would be nifty.

**Comment by ThibaultNocchi** [comment]
The server doesn't have such capabilities, so it'll have to be added there first :/

File: components/Buttons/MarkWatchedButton.vue (hunk 1 above)
Comment (Camc314) [reply components/Buttons/MarkWatchedButton.vue:2]: Without it, if you click on the icon, it redirect to the item details page instead of liking the item. Stop.prevent.... Is also on item menu and like button

File: components/Buttons/MarkWatchedButton.vue (hunk 2 above)
Comment (Camc314) [reply components/Buttons/MarkWatchedButton.vue:55]: I disagree, this is the same as the like button, and it makes more sense to set it after a success


I wonder what @ferferga thinks?

**Comment by Camc314** [comment]
> Can this be extended to music albums? Being able to filter out several thousand albums that I have heard or haven't heard would be nifty.

Alternatively you could like albums that you have listened two, and look for unliked ones in your library.

File: components/Buttons/MarkWatchedButton.vue (hunk 1 above)
Comment (ThibaultNocchi) [reply components/Buttons/MarkWatchedButton.vue:2]: Ok nice, though that's weird cause in my previous experiences with Vuetify I never did that

File: components/Buttons/MarkWatchedButton.vue (hunk 2 above)
Comment (ThibaultNocchi) [reply components/Buttons/MarkWatchedButton.vue:55]: I've never checked in that case, and I agree we need to be consistent. But then I'll say the same for the like button ;p

To me, a user should always have a direct reaction to something. If not, he'll click multiple times. So either we do this (setting `isPlayed` before the calls) or maybe we could replace the icon by a spinner during the request?

File: components/Buttons/MarkWatchedButton.vue [hunk 3]
```diff
@@ -0,0 +1,72 @@
+<template>
+  <v-btn v-if="canMarkWatched" icon dark @click.stop.prevent="togglePlayed">
+    <v-icon>
+      {{ isPlayed ? 'mdi-check-all' : 'mdi-check' }}
```
Comment (ThibaultNocchi) [inline components/Buttons/MarkWatchedButton.vue:4]: I'd not use this icon in the `isPlayed===true` state. I'd go with the same thing as jf-web, changing its color. The all-checked icon looks like we selected some stuff.

Maybe another thing, why not use an eye icon for the watch status instead of the check? (just an idea though)

**Review by ThibaultNocchi — COMMENTED** [review]
On the item page, there's no way to mark it as watched. I think it could go next to the liked icon

File: components/Buttons/MarkWatchedButton.vue (hunk 3 above)
Comment (Camc314) [reply components/Buttons/MarkWatchedButton.vue]: I wasn't a fan of the colors before. I've updated it to show blue if it has been watched, else white

File: components/Buttons/MarkPlayedButton.vue [hunk 4]
```diff
@@ -0,0 +1,70 @@
+<template>
+  <v-btn v-if="canMarkWatched" icon dark @click.stop.prevent="togglePlayed">
+    <v-icon :color="isPlayed ? 'blue' : 'white'"> mdi-check </v-icon>
```
Comment (ThibaultNocchi) [inline components/Buttons/MarkPlayedButton.vue:3]: I *think* you could use the `primary` color in that case, does it work?

Comment (Camc314) [reply components/Buttons/MarkPlayedButton.vue]: Good idea! Updated

**Review by ThibaultNocchi — COMMENTED** [review]
Seems all good, what about @ferferga advice on toggling the check icon before the query finishes?

**Comment by ferferga** [comment]
@ThibaultNocchi What do you mean exactly? 😅

I prefer @Perseusdehond icon choice here: https://github.com/jellyfin/jellyfin-vue/discussions/742. I think the tick one can be understood as a "multiple select" instead (in the library context).

I'm not 100% sure about it being in a different color in his designs. I like it but I have my reservations, what do you think?

**Comment by Camc314** [comment]
Imo the design in 742 looks more like a save for later instead of I've watched this or I haven't that's just my view.

In the future, with some help from server we could definitely add the watch later flag, like Perseusdehond suggested

**Comment by Camc314** [comment]
Current:
```
await marking played from server
set data to update icon
```

What ThibaultNocchi suggested:

```
set data to update icon
await marking played from server
```
The second one would provide a instant visual update, but may not work as well if it fails.

**Comment by Perseusdehond** [comment]
@ferferga  I'm not sure if you understood what the icon was for? It is for adding content to a watchlist, not for marking it played/unplayed. I don't mind it if icons get swapped, but I do agree with @Camc314. And btw I am looking forward to the day that feature is implemented as I suggested that it could already be done ;)

**Comment by ferferga** [comment]
Oh I see, this is what happens when reviewing stuff late :) @Perseusdehond 

Great idea, hope to get it added at some point soon, it will also depend on server changes, unless we do some client-side hackery, which I would avoid at all costs (better to do things right since the beginning).

@Camc314 I would 100% go for the second approach. If it fails, it's just a matter of doing ``this.bool = !this.bool``. It's one line and provide a much better UX.

File: components/Buttons/MarkPlayedButton.vue [hunk 5]
```diff
@@ -0,0 +1,70 @@
+<template>
+  <v-btn v-if="canMarkWatched" icon dark @click.stop.prevent="togglePlayed">
+    <v-icon :color="isPlayed ? 'primary' : 'white'"> mdi-check </v-icon>
```
Comment (ferferga) [inline components/Buttons/MarkPlayedButton.vue:3]:
```suggestion
    <v-icon :color="isPlayed ? 'primary' : 'white'">mdi-check</v-icon>
```

This isn't fixed automatically because the brace rule?

File: components/Buttons/MarkPlayedButton.vue [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        )
+      ) {
+        return true;
+      }
+
+      if (this.item.MediaType === 'Video' && this.item.Type !== 'TvChannel') {
+        return true;
+      }
+
+      return false;
```
Comment (ferferga) [inline components/Buttons/MarkPlayedButton.vue:38]: We have canPlay in a mixin for this. You can mix it with the check of ``item.UserData.Played`` existence if you want better "safety".

File: components/Buttons/MarkPlayedButton.vue [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        } else {
+          await this.$api.playState.markPlayedItem({
+            userId: this.$auth.user.Id,
+            itemId: this.item.Id || ''
+          });
+          this.isPlayed = true;
+        }
+      } catch (error) {
+        this.pushSnackbarMessage({
+          message: this.$t('unableToTogglePlayed'),
```
Comment (ferferga) [inline components/Buttons/MarkPlayedButton.vue:63]: Can we put all the snackbar messages under a ``snackbar`` subkey? Just do it for this one but let's make another PR to cleanup the rest of them.

File: components/Item/Card/Card.vue [hunk 8]
```diff
@@ -63,6 +63,7 @@
             v-if="overlay"
             class="card-lower-buttons d-flex justify-center align-center"
           >
+            <mark-played-button :item="item" />
```
Comment (ferferga) [inline components/Item/Card/Card.vue:66]: This needs dark so it's visible in light mode.

**Comment by Camc314** [comment]
Ok, Sounds good, I'll update it in the morning

I'll also make a PR for likeButton doing the same thing

**Review by ferferga — CHANGES_REQUESTED** [review]

File: components/Buttons/MarkPlayedButton.vue (hunk 6 above)
Comment (Camc314) [reply components/Buttons/MarkPlayedButton.vue:38]: The sets of canPlay and canBeLiked are slightly different

canLike
Type: Series, Season, BoxSet, AudioPodcast, AudioBook
MediaType: Video (Type !== TvChannel)

canPlay
Type: MusicGenre Season Series, BoxSet, MusicAlbum, MusicArtist, Playlist
MediaType: Video Audio

File: components/Buttons/MarkPlayedButton.vue (hunk 5 above)
Comment (Camc314) [reply components/Buttons/MarkPlayedButton.vue:3]: we seem to have a mix of both `> content<` and `>content<` throught the codebase

File: components/Buttons/MarkPlayedButton.vue (hunk 7 above)
Comment (Camc314) [reply components/Buttons/MarkPlayedButton.vue:63]: Sounds good, I'll do it in a seperate PR though

File: components/Item/Card/Card.vue (hunk 8 above)
Comment (Camc314) [reply components/Item/Card/Card.vue:66]: I'm forcing it to be dark in `MarkPlayedButton.vue` is there any point in adding a prop if it is dark everywhere?

File: components/Buttons/MarkWatchedButton.vue (hunk 2 above)
Comment (Camc314) [reply components/Buttons/MarkWatchedButton.vue]: Updated

File: components/Buttons/MarkPlayedButton.vue (hunk 6 above)
Comment (ferferga) [reply components/Buttons/MarkPlayedButton.vue:38]: I would place it in the mixin anyway, so it can be revisited later in case it's needed easily. Right now we have these kind of things scattered all around, which makes maintenance more difficult.

File: components/Item/Card/Card.vue (hunk 8 above)
Comment (ferferga) [reply components/Item/Card/Card.vue:66]: Then, in it's current status, it will look bad on itemdetails page.

File: components/Buttons/MarkPlayedButton.vue (hunk 6 above)
Comment (Camc314) [reply components/Buttons/MarkPlayedButton.vue:38]: I was thinking about this earlier, I'm probably going to try and collate all those kind of tools to one single file, I'll probably move it out the mixin to `itemHelper` since it doesn't really need to be a mixin

Comment (ferferga) [reply components/Buttons/MarkPlayedButton.vue:38]: Yeah, that's what I meant, no need for a new mixin, just add it to itemHelper.

File: components/Item/Card/Card.vue (hunk 8 above)
Comment (Camc314) [reply components/Item/Card/Card.vue]: Ah good catch, updated 👍

**Comment by Perseusdehond** [comment]
> Great idea, hope to get it added at some point soon, it will also depend on server changes, unless we do some client-side hackery, which I would avoid at all costs (better to do things right since the beginning).
> 
Yeah, my idea was displaying a playlist nicely as a watchlist although I don't know if this falls in the category of client-side hackery :)
But if it can be done, than do it the right way first, totally agree

File: components/Buttons/MarkPlayedButton.vue [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+  },
+  data() {
+    return {
+      isPlayed: false
+    };
+  },
+  computed: {
+    canMarkWatched(): boolean {
+      return canMarkWatched(this.item);
```
Comment (ferferga) [inline components/Buttons/MarkPlayedButton.vue:36]: Just move this to the template, no need for this computed property

**Review by ferferga — CHANGES_REQUESTED** [review]

**Review by ThibaultNocchi — APPROVED** [review]

**Review by MrTimscampi — COMMENTED** [review]
LGTM, but weirdly enough, it shows up on libraries too:

image

**Comment by ferferga** [comment]
Maybe we should check for ``canPlay`` and ``canMarkWatched``?

**Comment by Camc314** [comment]
> LGTM, but weirdly enough, it shows up on libraries too:



Should be fixed

**Review by MrTimscampi — APPROVED** [review]

**Review by ferferga — APPROVED** [review]
